### PR TITLE
[internal] Lint typescript, json, new eslint rules

### PR DESCRIPTION
**Pull request by kuceb**
Part of #2690 

the best way to view this PR diff is this filter: https://github.com/cypress-io/cypress/pull/4449/files?file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.yml&file-filters%5B%5D=dotfile&owned-by%5B%5D=
everything besides that is result of linting

motivation:
with typescript coming in, we need a way to lint it locally, in editor, and in ci. I wanted to use keep using eslint --fix, so we use `@ typescript-eslint`

changes:
- all the changes as a result of actually linting everything we want
- root-level `.eslintignore` that covers entire monorepo. This is good for editor integration (you'll get red squiggles when you need them, and won't when you don't) and running one `eslint` command from the root
- removes subpackages `lint` commands as the root now lints everything recursively (not the case for `coffeelint` since it has no ignore file)
- removes the need for `prettier` and `tslint` (expect for `cli/types` which requires tslint for `dtslint`)
- lints and auto-fixes .json or .eslintrc files, sorts package.json, with https://github.com/cypress-io/eslint-plugin-json
- use `lint-staged` properly to auto-fix partially staged files

#### tasks:
- [x] remove duplicate rules from `.eslintrc` that are in `cypress-dev/general` eslint rules
- [x] merge https://github.com/cypress-io/eslint-plugin-cypress-dev/pull/10 to modify  eslint rule, release new version
- [x] update source files with lint changes from https://github.com/cypress-io/eslint-plugin-cypress-dev/pull/10

- [x] use `@cypress/eslint-plugin-json` to enfore json style across project
- [x] lint all json
- [x] rename cypress-dev eslint plugin to `@cypress/dev`
- [x] add rule for allowing implicit braces only if oneline
- [x] wait for new version of @cypress/eslint-plugin-dev to release
- [x] add .vscode workspace settings for easier vscode setup

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -1,14 +1,42 @@
...
+      "error"
+    ],
+    "array-bracket-newline": [
+      "error",
+      "consistent"
+    ],
+    "arrow-body-style": [
+      "error",
+      "always"
+    ],
```
Comment (Bkucera) [inline .eslintrc]: formatted with default vscode JSON formatter

File: .eslintrc [hunk 2]
```diff
@@ -55,18 +83,49 @@
     "node": true
   },
   "parserOptions": {
+    "ecmaVersion": 2018,
```
Comment (Bkucera) [inline .eslintrc]: probably should bump this down to a lower version.

File: package.json [hunk 3]
```diff
@@ -33,11 +33,12 @@
...
     "clean-deps": "npm run all clean-deps && rm -rf node_modules",
     "docker": "./scripts/run-docker-local.sh",
-    "lint-js": "eslint --fix scripts/*.js packages/ts/*.js cli/*.js cli/**/*.js",
-    "lint-changed": "git diff --name-only | grep '\\.js$' | xargs npx eslint --fix",
-    "lint-coffee": "coffeelint scripts/**/*.coffee",
-    "lint": "npm run lint-js && npm run lint-coffee",
-    "pretest": "npm run lint && npm run all lint && npm run test-scripts",
+    "lint-js": "eslint --fix .",
+    "lint-ts": "eslint --fix . --ext .ts",
+    "lint-changed": "git diff HEAD --name-only | grep '\\.\\(ts\\|js\\)$' | xargs npx eslint --fix ''",
```
Comment (Bkucera) [inline package.json]: before, this was only linting `unstaged` files. using `git diff HEAD` includes both staged & unstaged

File: .eslintrc [hunk 4]
```diff
@@ -1,14 +1,42 @@
...
+    "@typescript-eslint"
+  ],
   "rules": {
-    "no-multiple-empty-lines": ["error", { "max": 1 } ],
-    "no-else-return": [ "error", { "allowElseIf": false } ],
-    "brace-style": ["error", "1tbs", { "allowSingleLine": false }],
-    "no-unneeded-ternary": ["error"],
-    "array-bracket-newline": ["error", "consistent"],
-    "arrow-body-style": ["error", "always"],
+    "no-multiple-empty-lines": [
```
Comment (chrisbreiding) [inline .eslintrc]: I think these rules can be cleaned up and simplified. A lot of them have moved into the [`cypress-dev/general` plugin](https://github.com/cypress-io/eslint-plugin-cypress-dev/blob/master/index.js), so any overlap can be removed from here.

File: packages/network/test/.eslintrc [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (chrisbreiding) [inline packages/network/test/.eslintrc]: What's up with the huge amount of spacing?

Comment (Bkucera) [reply packages/network/test/.eslintrc]: I'm using vscode's default JSON formatter, for consistency

File: .eslintrc (hunk 2 above)
Comment (chrisbreiding) [reply .eslintrc]: Why bump it to a lower version? I think we want the latest version or the latest features won't work. I believe a higher version includes support for all lower versions.

Comment (Bkucera) [reply .eslintrc]: @chrisbreiding agreed

File: .eslintrc (hunk 4 above)
Comment (Bkucera) [reply .eslintrc]: ok, I removed all duplicates :+1:

File: .eslintignore [hunk 6]
```diff
@@ -1,6 +1,21 @@
...
+!packages/example/lib
 packages/server/lib/scaffold/plugins/index.js
 packages/server/lib/scaffold/support/index.js
 packages/server/lib/scaffold/support/commands.js
-packages/server/test/fixtures
-packages/example/cypress
+packages/extension/test/helpers/background.js
+packages/network/lib/*.js
+
+# these generated files are co-located with .ts source files
```
Comment (brian-mann) [inline .eslintignore]: these can be removed now

File: .eslintignore [hunk 7]
```diff
@@ -1,6 +1,21 @@
...
+packages/example
+cli/types
+!packages/example/lib
 packages/server/lib/scaffold/plugins/index.js
 packages/server/lib/scaffold/support/index.js
 packages/server/lib/scaffold/support/commands.js
-packages/server/test/fixtures
-packages/example/cypress
+packages/extension/test/helpers/background.js
+packages/network/lib/*.js
```
Comment (brian-mann) [inline .eslintignore]: probably this too

File: cli/schema/cypress.schema.json [hunk 8]
```diff
@@ -157,7 +175,7 @@
       "default": 1000,
       "description": "Default width in pixels for the application under tests’ viewport. (Override with cy.viewport() command)"
     },
-    "animationDistanceThreshold":	{
+    "animationDistanceThreshold": {
       "type": "number",
       "default": 5,
       "description": "The distance in pixels an element must exceed over time to be considered animating"
```
Comment (Bkucera) [inline cli/schema/cypress.schema.json]: formatted with vscode default JSON formatter

File: package.json [hunk 9]
```diff
@@ -62,17 +60,23 @@
     "effective:circle:config": "circleci config process circle.yml | sed /^#/d"
   },
   "lint-staged": {
-    "*.js": [
-      "npm run precommit-lint"
+    "*.(js|ts)": [
+      "eslint --fix"
```
Comment (brian-mann) [inline package.json]: move this to an npm command

**Review by brian-mann — CHANGES_REQUESTED** [review]
let's add a new `scripts/eslint-verify-and-fix.js` and hook it into circle and `npm` that both verifies and attempts to `--fix`.

What this is solving is situations where our base `.eslintrc` rules change - but then on forked PR's that have diverge - those get "missed".

We want to error during the linting stage, but we also want to provide a really good error message telling the author exactly what went wrong and how to fix it locally and push changes.

IDEA: rather than let the user do this, we could alternatively just use a `cypress-bot` to automatically push these fixes if we're on a PR.

**Comment by brian-mann** [comment]
Other areas worth exploring...

- sorting `package.json` pre-commit or pre-push
- sorting npm `require` or `imports` 

Worth verifying...

- what happens on `precommit` hooks when there are staged files but only certain lines of the file staged?

**Comment by brian-mann** [comment]
Also let's...

- [ ] Add a JSON beautifier plugin to `eslint`. Something like https://github.com/lydell/json-stringify-pretty-compact or https://github.com/AitoDotAI/json-stringify-pretty-compact
- [ ] Create another plugin that allows implicit returns (and no curly braces) but only if it's a single line, not multiline. (until we do this, please add back in the eslint rule you removed)

**Comment by brian-mann** [comment]
@Bkucera 	can we disable sourcemaps when building cypress for prod - per our last discussion? That way the `eval` sourcemaps only affect development.

**Review by brian-mann — CHANGES_REQUESTED** [review]
This PR is accidentally(?) reverting work done in `develop` such as changes to the `cli/schema/cypress.schema.json` found here: https://github.com/cypress-io/cypress/pull/4449/files#diff-4e080e63eb738f2552ed87596970c631

The PR went in a few days ago here: https://github.com/cypress-io/cypress/pull/4300

Why is this happening? What other PR's is this PR overriding?

**Comment by Bkucera** [comment]
@brian-mann my mistake. when i went to re-lint all the files, I git reset to origin/develop without merging it first, then assumed the json files were up to date with develop.

Fixed now

**Comment by Bkucera** [comment]
wow, appveyor.yml thought i was trying to background a process with the '&' in my commit message.
Fixed now

**Pull request merged**